apps/aevm/lib/gas.ex

```diff
@@ -0,0 +1,156 @@
+defmodule Gas do
+  @moduledoc """
+    Module for calculating the gas of a contract
+  """
+
+  use Bitwise
+
+  require OpCodesUtil
+  require GasCodes
+
+  @spec update_gas(integer(), map()) :: map() | {:error, String.t(), map()}
+  def update_gas(gas_cost, state) do
+    curr_gas = State.gas(state)
+
+    if curr_gas >= gas_cost do
+      gas_after = curr_gas - gas_cost
+      State.set_gas(gas_after, state)
+    else
+      throw({:error, "out_of_gas", state})
+    end
+  end
+
+  @spec op_gas_cost(char()) :: integer()
+  def op_gas_cost(op_code) do
+    {_name, _pushed, _popped, op_gas_price} = OpCodesUtil.opcode(op_code)
+    op_gas_price
+  end
+
+  @spec memory_gas_cost(map(), map()) :: integer()
+  def memory_gas_cost(state_with_ops, state_without) do
+    words1 = Memory.memory_size_words(state_with_ops)
+
+    case Memory.memory_size_words(state_without) do
+      ^words1 ->
+        0
+
+      words2 ->
+        first = round(GasCodes._GMEMORY() * words1 + Float.floor(words1 * words1 / 512))
+        second = round(GasCodes._GMEMORY() * words2 + Float.floor(words2 * words2 / 512))
+        first - second
+    end
+  end
+
+  @spec dynamic_gas_cost(String.t(), map()) :: integer()
+  def dynamic_gas_cost("CALL", state) do
+    # TODO: account creation?
+
+    gas_cost = GasCodes._GCALL()
+
+    gas_state = State.gas(state)
+    gas = Stack.peek(0, state)
+    value = Stack.peek(2, state)
+
+    gas_cost =
+      gas_cost +
+        if value !== 0 do
+          GasCodes._GCALLVALUE()
+        else
+          0
+        end
+
+    gas_cost +
+      if gas_state >= gas_cost do
+        gas_one_64_substracted = substract_one_64(gas_state - gas_cost)
+
+        if gas < gas_one_64_substracted do
+          gas
+        else
+          gas_one_64_substracted
+        end
+      else
+        gas
+      end
+  end
+
+  def dynamic_gas_cost("DELEGATECALL", state) do
+    # TODO
```

Comment:
Forgotten `TODO`?

Reply:
No, it will be implemented in an iteration to come